nvflare/lighter/utils.py

```diff
@@ -12,21 +12,109 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import json
 import os
 import random
-from base64 import b64encode
+import shutil
+from base64 import b64decode, b64encode
 
 import yaml
-from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes, serialization
 from cryptography.hazmat.primitives.asymmetric import padding
 
+from nvflare.lighter.impl.cert import load_crt
+
 
 def generate_password(passlen=16):
     s = "abcdefghijklmnopqrstuvwxyz01234567890ABCDEFGHIJKLMNOPQRSTUVWXYZ"
     p = "".join(random.sample(s, passlen))
     return p
 
 
+def sign_one(content, signing_pri_key):
+    signature = signing_pri_key.sign(
+        data=content,
+        padding=padding.PSS(
+            mgf=padding.MGF1(hashes.SHA256()),
+            salt_length=padding.PSS.MAX_LENGTH,
+        ),
+        algorithm=hashes.SHA256(),
+    )
+    return b64encode(signature).decode("utf-8")
+
+
+def load_private_key_file(file_path):
+    with open(file_path, "rt") as f:
+        pri_key = serialization.load_pem_private_key(f.read().encode("ascii"), password=None, backend=default_backend())
+    return pri_key
+
+
+def sign_folders(folder, signing_pri_key, crt_path):
+    for root, folders, files in os.walk(folder):
+        signatures = dict()
+        for file in files:
+            if file == ".__nvfl_sig.json":
+                continue
+            signature = signing_pri_key.sign(
+                data=open(os.path.join(root, file), "rb").read(),
+                padding=padding.PSS(
+                    mgf=padding.MGF1(hashes.SHA256()),
+                    salt_length=padding.PSS.MAX_LENGTH,
+                ),
+                algorithm=hashes.SHA256(),
+            )
+            signatures[file] = b64encode(signature).decode("utf-8")
+        for folder in folders:
+            signature = signing_pri_key.sign(
+                data=folder.encode("utf-8"),
```

Comment:
Should we recursively sign the files content within a folder? or simply sign the folder.encode("utf-8") data?

Reply:
The os.walk does the recursive traversal into each folder.  The content of the folder, ie files and subfolders, are signed and the signature file is inside the folder.  So files are signed by contents and subfolders are signed by folder names.